aws/resource_aws_ec2_traffic_mirror_filter_rule.go

```diff
@@ -0,0 +1,370 @@
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"accept",
+					"reject",
+				}, false),
+			},
+			"rule_number": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"source_cidr_block": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateCIDRNetworkAddress,
+			},
+			"source_port_range": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"from_port": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"to_port": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"traffic_direction": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"ingress",
+					"egress",
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceAwsEc2TrafficMirrorFilterRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	filterId := d.Get("traffic_mirror_filter_id")
+
+	input := &ec2.CreateTrafficMirrorFilterRuleInput{
+		TrafficMirrorFilterId: aws.String(filterId.(string)),
+		DestinationCidrBlock:  aws.String(d.Get("destination_cidr_block").(string)),
+		SourceCidrBlock:       aws.String(d.Get("source_cidr_block").(string)),
+		RuleAction:            aws.String(d.Get("rule_action").(string)),
+		RuleNumber:            aws.Int64(int64(d.Get("rule_number").(int))),
+		TrafficDirection:      aws.String(d.Get("traffic_direction").(string)),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("protocol"); ok {
+		input.Protocol = aws.Int64(int64(v.(int)))
+	}
+
+	if v, ok := d.GetOk("destination_port_range"); ok {
+		input.SetDestinationPortRange(buildTrafficMirrorPortRangeRequest(v.([]interface{})))
+	}
+
+	if v, ok := d.GetOk("source_port_range"); ok {
+		input.SetSourcePortRange(buildTrafficMirrorPortRangeRequest(v.([]interface{})))
+	}
+
+	out, err := conn.CreateTrafficMirrorFilterRule(input)
+	if err != nil {
+		return fmt.Errorf("Error creating traffic mirror filter rule for %v", filterId)
+	}
+
+	d.SetId(*out.TrafficMirrorFilterRule.TrafficMirrorFilterRuleId)
+	return resourceAwsEc2TrafficMirrorFilterRuleRead(d, meta)
+}
+
+func resourceAwsEc2TrafficMirrorFilterRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+	ruleId := d.Id()
+
+	var rule *ec2.TrafficMirrorFilterRule
+	filterId, filterIdSet := d.GetOk("traffic_mirror_filter_id")
+	input := &ec2.DescribeTrafficMirrorFiltersInput{}
+	if filterIdSet {
+		input.TrafficMirrorFilterIds = []*string{aws.String(filterId.(string))}
+		awsMutexKV.Lock(filterId.(string))
+		defer awsMutexKV.Unlock(filterId.(string))
+	} else {
+		awsMutexKV.Lock("DescribeTrafficMirrorFilters")
+		defer awsMutexKV.Unlock("DescribeTrafficMirrorFilters")
+	}
```

Comment:
`traffic_mirror_filter_id` is always set since its required, however this logic to use mutexes during read is very uncommon, can you explain why its included here (potentially adding a code comment)? If its not necessary, then the logic can be simplified:

```go
	input := &ec2.DescribeTrafficMirrorFiltersInput{
		TrafficMirrorFilterIds: aws.StringSlice([]string{d.Get("traffic_mirror_filter_id").(string)}),
	}
```

Reply:
What about during import?

Reply:
If import requires special logic, would suggest moving to the import function. Otherwise, the more preferred solution is to have all necessary identifiers in the resource import ID.